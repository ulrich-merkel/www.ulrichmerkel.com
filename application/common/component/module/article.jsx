/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import ArticleHeadline from './article/headline';
import ArticleLead from './article/lead';
import ArticleButton from './article/button';
import { Meta } from '../element/meta';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export function ModuleArticle(props) {
    const {
        componentType,
        className,
        isDialog,
        itemType,
        isMain,
        isSpaced,
        noMargin,
        content,
        children,
        ...otherProps
    } = props;

    const ComponentType = componentType;
    const componentClassName = classnames(
        {
            'is-main': isMain,
            'has-no-margin': noMargin
        },
        'm-article',
        className
    );
    const composedHeadlineClassName = classnames(
        {
            'is-spaced': isSpaced
        },
        'm-article__headline'
    );

    return (
        <ComponentType
            className={componentClassName}
            itemScope
            itemType={itemType}
            {...otherProps}
        >
            <ArticleHeadline
                className={composedHeadlineClassName}
                text={content.headline}
                {...{ isMain }}
            />
            <ArticleLead className={'m-article__lead'} text={content.lead} />
            <div className="m-article__text" itemProp="text">
                {children}
            </div>
            <ArticleButton
                btnTo={content.btnTo}
                btnLabel={content.btnLabel}
                btnTitle={content.btnTitle}
                className={'m-article__button'}
                {...{ isDialog }}
            />
            <Meta itemProp="name" content={content.headline} />
            <Meta
                itemProp="author"
                content={content.author}
                defaultContent="Ulrich Merkel"
            />
            <Meta itemProp="datePublished" content={content.datePublished} />
        </ComponentType>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [componentType='article'] - The component element type used for React.createElement
 * @property {string} [className] - The component css class names, will be merged into component default classNames
 * @property {string} [isDialog] -
 * @property {boolean} [isMain=false] - Whether the component text should be displayed as main article or not
 * @property {boolean} [noMargin=false] - Whether the component has no css margin or not
 * @property {boolean} [isSpaced=false] - Whether the component headline has a spaced grid or not
 * @property {string} [itemType='https://schema.org/Article'] - The schema.org itemtype url attribute
 * @property {Array|string} [children] - The component dom node childs, usally an array of components, if there is only a single child it's a string
 * @property {object} [content={}] - The component translation config
 */
ModuleArticle.propTypes = {
    componentType: PropTypes.string,
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    isDialog: PropTypes.bool,
    isMain: PropTypes.bool,
    noMargin: PropTypes.bool,
    isSpaced: PropTypes.bool,
    itemType: PropTypes.string,
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.shape({
        btnTo: PropTypes.string,
        btnLabel: PropTypes.string,
        btnTitle: PropTypes.string,
        headline: PropTypes.string,
        lead: PropTypes.string,
        datePublished: PropTypes.string,
        author: PropTypes.string
    })
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleArticle.propTypes
 */
ModuleArticle.defaultProps = {
    isDialog: false,
    componentType: 'article',
    itemType: 'https://schema.org/Article',
    isSpaced: false,
    isMain: false,
    noMargin: false,
    content: {}
};
