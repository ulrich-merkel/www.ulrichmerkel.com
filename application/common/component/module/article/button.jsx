/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires common/component/element/paragraph
 * @requires common/component/element/button
 *
 * @changelog
 * - 0.0.4 Restructed module as simple wrapper article (rendering without children behaviour)
 * - 0.0.3 Moved to stateless function
 * - 0.0.2 Rewritten for es2015
 * - 0.0.1 Basic functions and structure
 *
 * @example <caption>Example usage (jsx)</caption>
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import {
    Button,
    P
} from '../../element';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {Object} [props] - The current component props
 * @param {string} [props.btnLabel=''] - The button label
 * @param {string} [props.btnTitle=''] - The button title
 * @param {string} [props.btnTo=''] - The button link target
 * @param {string} [props.className] - The component css class names, will be merged into component default classNames
 * @param {boolean} [props.isDialog=false] - Flag if this is rendering within a dialog
 * @returns {ReactElement} React component markup
 */
function ModuleArticleButton(props) {
    const {
        btnLabel,
        btnTitle,
        btnTo,
        className,
        isDialog
    } = props;

    if (!btnTo || !btnLabel || isDialog) {
        return null;
    }

    const composedClassName = classnames(className, 'hide-on-print');

    return (
        <P className={composedClassName} isCentered>
            <Button title={btnTitle || btnLabel} to={btnTo}>
                {btnLabel}
            </Button>
        </P>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 */
ModuleArticleButton.propTypes = {
    btnLabel: PropTypes.string,
    btnTitle: PropTypes.string,
    btnTo: PropTypes.string,
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    isDialog: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {Object}
 */
ModuleArticleButton.defaultProps = {
    btnLabel: '',
    btnTitle: '',
    btnTo: '',
    isDialog: false
};

export default ModuleArticleButton;
