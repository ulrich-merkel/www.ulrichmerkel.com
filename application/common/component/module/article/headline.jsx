/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires common/component/element/headline
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Headline from '../../element/headline';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {Object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function ModuleArticleHeadline(props) {

    const {
        text,
        isMain,
        className
    } = props;

    if (!text) {
        return null;
    }

    const composedClassName = classnames(className);
    const htmlElement = isMain ? 'h1' : 'h2';

    return (
        <Headline className={composedClassName} itemProp='headline' htmlElement={htmlElement}>
            {text}
        </Headline>
    );

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {string} [text=''] - The headline text
 * @property {boolean} [isMain=false] - Whether this is a main headline (h1) or not
 * @property {string} [className] - The component css class names, will be merged into component default classNames
 */
ModuleArticleHeadline.propTypes = {
    text: PropTypes.string,
    isMain: PropTypes.bool,
    className: PropTypes.string // eslint-disable-line react/require-default-props
};

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @see ModuleArticleHeadline.propTypes
 */
ModuleArticleHeadline.defaultProps = {
    text: '',
    isMain: false
};

export default ModuleArticleHeadline;
