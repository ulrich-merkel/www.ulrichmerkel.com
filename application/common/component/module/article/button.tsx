/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';

import { Button } from '../../element/button';
import { P } from '../../element/paragraph';

type Props = {
    btnLabel?: string;
    btnTitle?: string;
    btnTo?: string;
    className?: string;
    isDialog?: boolean;
}

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @param {string} [props.btnLabel=''] - The button label
 * @param {string} [props.btnTitle=''] - The button title
 * @param {string} [props.btnTo=''] - The button link target
 * @param {string} [props.className] - The component css class names, will be merged into component default classNames
 * @param {boolean} [props.isDialog=false] - Flag if this is rendering within a dialog
 * @returns {ReactElement} React component markup
 */
export const ModuleArticleButton: FunctionComponent<Props> = (props) => {
    const { btnLabel = '', btnTitle = '', btnTo = '', className, isDialog = false } = props;

    if (!btnTo || !btnLabel || isDialog) {
        return null;
    }

    const composedClassName = classnames(className, 'hide-on-print');

    return (
        <P className={composedClassName} isCentered>
            <Button title={btnTitle || btnLabel} to={btnTo}>
                {btnLabel}
            </Button>
        </P>
    );
}
