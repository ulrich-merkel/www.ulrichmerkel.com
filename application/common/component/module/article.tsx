/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';

import { ModuleArticleHeadline } from './article/headline';
import { ModuleArticleLead } from './article/lead';
import { ModuleArticleButton } from './article/button';
import { Meta } from '../element/meta';

type Props = {
    children?: ReactNode;
    className?: string;
    content?: {
        btnTo: string;
        btnLabel: string;
        btnTitle: string;
        headline: string;
        lead: string;
        datePublished: string;
        author: string;
    };
    htmlElement?: keyof JSX.IntrinsicElements;
    isDialog?: boolean;
    isMain?: boolean;
    isSpaced?: boolean;
    itemType?: string;
    noMargin?: boolean;
}

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleArticle: FunctionComponent<Props> = (props) => {
    const {
        children,
        className,
        htmlElement: HtmlElement = 'article',
        content = {},
        isDialog = false,
        isMain = false,
        isSpaced = false,
        itemType = 'https://schema.org/Article',
        noMargin = false,
        ...otherProps
    } = props;

    const componentClassName = classnames(
        {
            'is-main': isMain,
            'has-no-margin': noMargin
        },
        'm-article',
        className
    );
    const composedHeadlineClassName = classnames(
        {
            'is-spaced': isSpaced
        },
        'm-article__headline'
    );

    return (
        <HtmlElement
            className={componentClassName}
            itemScope
            itemType={itemType}
            {...otherProps}
        >
            <ModuleArticleHeadline
                className={composedHeadlineClassName}
                text={content.headline}
                {...{ isMain }}
            />
            <ModuleArticleLead
                className={'m-article__lead'}
                text={content.lead}
            />
            <div className="m-article__text" itemProp="text">
                {children}
            </div>
            <ModuleArticleButton
                btnTo={content.btnTo}
                btnLabel={content.btnLabel}
                btnTitle={content.btnTitle}
                className={'m-article__button'}
                {...{ isDialog }}
            />
            <Meta itemProp="name" content={content.headline} />
            <Meta
                itemProp="author"
                content={content.author}
                defaultContent="Ulrich Merkel"
            />
            <Meta itemProp="datePublished" content={content.datePublished} />
        </HtmlElement>
    );
}
