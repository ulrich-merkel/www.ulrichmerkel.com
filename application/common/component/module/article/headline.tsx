/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';

import { Headline } from '../../element/headline';

type Props = {
    text?: string;
    isMain?: boolean;
    className?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleArticleHeadline: FunctionComponent<Props> = (props) => {
    const { text = '', isMain = false, className } = props;

    if (!text) {
        return null;
    }

    const composedClassName = classnames(className);
    const htmlElement = isMain ? 'h1' : 'h2';

    return (
        <Headline
            className={composedClassName}
            itemProp="headline"
            htmlElement={htmlElement}
        >
            {text}
        </Headline>
    );
};
