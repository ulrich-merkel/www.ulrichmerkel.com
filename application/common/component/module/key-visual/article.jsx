/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires common/component/element/headline
 * @requires common/component/element/paragraph
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 *
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Headline } from '../../element/headline';
import { P } from '../../element/paragraph';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function ModuleKeyVisualArticle(props) {
    const { headline, lead } = props;

    if (!headline || !lead) {
        return null;
    }

    const componentArticleClassName = classnames('m-key-visual__text');

    return (
        <article className={componentArticleClassName}>
            <Headline
                className="m-key-visual__headline"
                itemProp="headline"
                htmlElement="h3"
            >
                <span className="m-key-visual__boxed">{headline}</span>
            </Headline>
            <P className="m-key-visual__lead" itemProp="description">
                <span className="m-key-visual__boxed">{lead}</span>
            </P>
        </article>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [headline=''] - The article headline content
 * @property {string} [lead=''] - The article lead content
 */
ModuleKeyVisualArticle.propTypes = {
    headline: PropTypes.string,
    lead: PropTypes.string
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleKeyVisualArticle.propTypes
 */
ModuleKeyVisualArticle.defaultProps = {
    headline: '',
    lead: ''
};

export default ModuleKeyVisualArticle;
