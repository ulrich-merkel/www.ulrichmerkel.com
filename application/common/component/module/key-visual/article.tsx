/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';

import { Headline } from '../../element/headline';
import { P } from '../../element/paragraph';
import { Article } from '../../element/article';

type Props = {
    headline?: string;
    lead?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleKeyVisualArticle: FunctionComponent<Props> = (props) => {
    const { headline, lead } = props;

    if (!headline || !lead) {
        return null;
    }

    const componentArticleClassName = classnames('m-key-visual__text');

    return (
        <Article className={componentArticleClassName}>
            <Headline
                className="m-key-visual__headline"
                itemProp="headline"
                htmlElement="h3"
            >
                <span className="m-key-visual__boxed">{headline}</span>
            </Headline>
            <P className="m-key-visual__lead" itemProp="description">
                <span className="m-key-visual__boxed">{lead}</span>
            </P>
        </Article>
    );
};
