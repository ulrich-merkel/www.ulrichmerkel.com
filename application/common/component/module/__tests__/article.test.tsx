import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { ModuleArticle } from '../article';

describe('ModuleArticle', function fnDescribe() {
    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <ModuleArticle
                htmlElement="span"
                className="module-article"
                isMain
                noMargin
                isSpaced
                content={{
                    headline: 'Headline Article',
                    lead: 'Lead Article',
                    datePublished: '20161212',
                    author: 'Author'
                }}
            >
                Module article children
            </ModuleArticle>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
