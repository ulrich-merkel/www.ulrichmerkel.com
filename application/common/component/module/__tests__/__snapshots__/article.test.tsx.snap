// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleArticle should render correctly 1`] = `
<DocumentFragment>
  <article
    class="c-type--article is-main has-no-margin m-article module-article"
    itemscope=""
    itemtype="https://schema.org/Article"
  >
    <h1
      class="is-centered c-type--h1 is-spaced m-article__headline"
      itemprop="headline"
    >
      Headline Article
    </h1>
    <p
      class="is-centered c-type--p m-article__lead"
      itemprop="alternativeHeadline"
    >
      <strong>
        Lead Article
      </strong>
    </p>
    <div
      class="m-article__text"
      itemprop="text"
    >
      Module article children
    </div>
    <meta
      content="Headline Article"
      itemprop="name"
    />
    <meta
      content="Author"
      itemprop="author"
    />
    <meta
      content="20161212"
      itemprop="datePublished"
    />
  </article>
</DocumentFragment>
`;
